horovod/common/operations.cc

```diff
@@ -840,18 +847,147 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+    // allgather into the shared buffer first, then memcpy to output buffer
+    if (horovod_global.is_homogeneous) {
+      int element_size;
+      MPI_Type_size(GetMPIDataType(e.tensor), &element_size);
+
+      auto* cross_recvcounts = new int[horovod_global.cross_size]();
+      auto* cross_displcmnts = new int[horovod_global.cross_size]();
+      for (int i = 0; i < horovod_global.cross_size; i++) {
+        cross_recvcounts[i] = recvcounts[horovod_global.local_size * i +
+                                         horovod_global.local_rank];
+        cross_displcmnts[i] = displcmnts[horovod_global.local_size * i +
+                                         horovod_global.local_rank];
+      }
+      int64_t total_size = recvcounts[tensor_sizes.size() - 1] +
+                           displcmnts[tensor_sizes.size() - 1];
+
+      // If shared buffer is not initialized or is not large enough, reallocate
+      if (horovod_global.shared_buffer == nullptr ||
+          horovod_global.shared_buffer_size < total_size * element_size) {
+        if (horovod_global.shared_buffer != nullptr) {
+          MPI_Win_fence(0, horovod_global.window);
+          MPI_Win_free(&horovod_global.window);
+          horovod_global.shared_buffer = nullptr;
+        }
+        int64_t window_size =
+            horovod_global.local_rank == 0 ? total_size * element_size : 0;
+
+        // Allocate shared memory, give each rank their respective pointer
+        ACTIVITY_START_ALL(entries, timeline, ALLOCATE_SHARED_BUFFER)
+        MPI_Win_allocate_shared(
+            window_size, element_size, MPI_INFO_NULL, horovod_global.local_comm,
+            &horovod_global.shared_buffer, &horovod_global.window);
+
+        if (horovod_global.local_rank != 0) {
+          int disp_unit;
+          MPI_Aint winsize;
+          MPI_Win_shared_query(horovod_global.window, 0, &winsize, &disp_unit,
+                               &horovod_global.shared_buffer);
+        }
+        MPI_Win_fence(0, horovod_global.window);
+        horovod_global.shared_buffer_size = total_size * element_size;
+        ACTIVITY_END_ALL(entries, timeline)
+      }
+#if HAVE_CUDA
+      bool on_gpu = e.device != CPU_DEVICE_ID;
+      if (on_gpu) {
+        cudaStream_t& stream = horovod_global.streams[e.device];
+
+        CUDA_CHECK(entries, "cudaSetDevice", cudaSetDevice(e.device))
+
+        // Ensure stream is in the map before executing reduction.
+        if (stream == nullptr) {
+          int greatest_priority;
+          CUDA_CHECK(entries, "cudaDeviceGetStreamPriorityRange",
+                     cudaDeviceGetStreamPriorityRange(NULL, &greatest_priority))
+          CUDA_CHECK(entries, "cudaStreamCreateWithPriority",
+                     cudaStreamCreateWithPriority(
+                         &stream, cudaStreamNonBlocking, greatest_priority))
+        }
+        auto event_queue = std::queue<std::pair<std::string, cudaEvent_t>>();
```

Comment:
Are you using this event queue?

Reply:
I think you need to add record event & WAIT_FOR_EVENTS to synchronize before MPI_Allgatherv

Reply:
I added `RECORD_EVENT` and `WAIT_FOR_EVENTS` after memcpy operations.